### PR TITLE
gh-108082: Use PyErr_FormatUnraisable()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Replace most of calls of _PyErr_WriteUnraisableMsg() and some calls of PyErr_WriteUnraisable(NULL) with PyErr_FormatUnraisable().

* Issue: gh-108082

### CODE REVIEW

File: Modules/gcmodule.c [hunk 1]
```diff
@@ -1032,8 +1032,8 @@ delete_garbage(PyThreadState *tstate, GCState *gcstate,
                 Py_INCREF(op);
                 (void) clear(op);
                 if (_PyErr_Occurred(tstate)) {
-                    _PyErr_WriteUnraisableMsg("in tp_clear of",
-                                              (PyObject*)Py_TYPE(op));
+                    PyErr_FormatUnraisable("Exception ignored in tp_clear of %.200s",
```
Comment (vstinner) [inline Modules/gcmodule.c]: IMO `%.200s` is a bug here. We should no longer truncate type names if they are too long. Many years ago, Python had an internal buffer of a fixed size and so the overall string must not be too long. But this has been fixed.

Please don't truncate type names.

File: Modules/getpath.c [hunk 2]
```diff
@@ -911,7 +910,7 @@ _PyConfig_InitPathConfig(PyConfig *config, int compute_path_config)
     ) {
         Py_DECREF(co);
         Py_DECREF(dict);
-        _PyErr_WriteUnraisableMsg("error evaluating initial values", NULL);
+        PyErr_FormatUnraisable("Exception ignored in evaluating initial values");
```
Comment (vstinner) [inline Modules/getpath.c]:
```suggestion
        PyErr_FormatUnraisable("Exception ignored in preparing getpath");
```

File: Modules/getpath.c [hunk 3]
```diff
@@ -920,13 +919,13 @@ _PyConfig_InitPathConfig(PyConfig *config, int compute_path_config)
 
     if (!r) {
         Py_DECREF(dict);
-        _PyErr_WriteUnraisableMsg("error evaluating path", NULL);
+        PyErr_FormatUnraisable("Exception ignored in evaluating path");
```
Comment (vstinner) [inline Modules/getpath.c]:
```suggestion
        PyErr_FormatUnraisable("Exception ignored in running getpath");
```

File: Objects/moduleobject.c [hunk 4]
```diff
@@ -902,10 +902,9 @@ module_clear(PyModuleObject *m)
     {
         int res = m->md_def->m_clear((PyObject*)m);
         if (PyErr_Occurred()) {
-            PySys_FormatStderr("Exception ignored in m_clear of module%s%V\n",
-                               m->md_name ? " " : "",
-                               m->md_name, "");
-            PyErr_WriteUnraisable(NULL);
+            PyErr_FormatUnraisable("Exception ignored in m_clear of module%s%V\n",
+                                   m->md_name ? " " : "",
+                                   m->md_name, "");
```
Comment (vstinner) [inline Objects/moduleobject.c]:
```suggestion
            PyErr_FormatUnraisable("Exception ignored in m_clear of module%s%V",
                                   m->md_name ? " " : "",
                                   m->md_name);
```

File: Objects/typeobject.c [hunk 5]
```diff
@@ -828,7 +828,7 @@ PyType_Modified(PyTypeObject *type)
             if (bits & 1) {
                 PyType_WatchCallback cb = interp->type_watchers[i];
                 if (cb && (cb(type) < 0)) {
-                    PyErr_WriteUnraisable((PyObject *)type);
+                    PyErr_FormatUnraisable("Exception ignored in watcher callback for %R", i, type);
```
Comment (vstinner) [inline Objects/typeobject.c]: If you want to log *i* variable, you should add ``%i`` somewhere.

File: Python/import.c [hunk 6]
```diff
@@ -584,7 +584,7 @@ _PyImport_ClearModulesByIndex(PyInterpreterState *interp)
     if (PyList_SetSlice(MODULES_BY_INDEX(interp),
                         0, PyList_GET_SIZE(MODULES_BY_INDEX(interp)),
                         NULL)) {
-        PyErr_WriteUnraisable(MODULES_BY_INDEX(interp));
+        PyErr_FormatUnraisable("Exception ignored on clearing interpreters module-list");
```
Comment (vstinner) [inline Python/import.c]:
```suggestion
        PyErr_FormatUnraisable("Exception ignored on clearing interpreters module list");
```

File: Python/perf_trampoline.c [hunk 7]
```diff
@@ -277,8 +274,7 @@ new_code_arena(void)
     if (new_arena == NULL) {
         PyErr_NoMemory();
         munmap(memory, mem_size);
-        _PyErr_WriteUnraisableMsg("Failed to allocate new code arena struct",
-                                  NULL);
+        PyErr_FormatUnraisable("Failed to allocate new code arena struct");
```
Comment (vstinner) [inline Python/perf_trampoline.c]:
```suggestion
        PyErr_FormatUnraisable("Failed to allocate new code arena struct for perf trampoline");
```

File: Objects/moduleobject.c [hunk 8]
```diff
@@ -902,10 +902,9 @@ module_clear(PyModuleObject *m)
     {
         int res = m->md_def->m_clear((PyObject*)m);
         if (PyErr_Occurred()) {
-            PySys_FormatStderr("Exception ignored in m_clear of module%s%V\n",
-                               m->md_name ? " " : "",
-                               m->md_name, "");
-            PyErr_WriteUnraisable(NULL);
+            PyErr_FormatUnraisable("Exception ignored in m_clear of module%s%V",
+                                   m->md_name ? " " : "",
+                                   m->md_name, "");
```
Comment (vstinner) [inline Objects/moduleobject.c:907]: The last ``""`` is not used, it can be removed.

File: Objects/typeobject.c [hunk 9]
```diff
@@ -828,7 +828,9 @@ PyType_Modified(PyTypeObject *type)
             if (bits & 1) {
                 PyType_WatchCallback cb = interp->type_watchers[i];
                 if (cb && (cb(type) < 0)) {
-                    PyErr_WriteUnraisable((PyObject *)type);
+                    PyErr_FormatUnraisable(
+                        "Exception ignored in watcher callback #%d for %R",
```
Comment (vstinner) [inline Objects/typeobject.c]:
```suggestion
                        "Exception ignored in type watcher callback #%d for %R",
```

**Review by vstinner — APPROVED** [review]
LGTM. I just left two minor suggestions.

**Comment by vstinner** [comment]
I wanted to do this change for many years. It's awful to get a warning without *any* context :-( Thanks for this nice enhancement! I like PyErr_FormatUnraisable() API!

**Comment by serhiy-storchaka** [comment]
We both wanted it. But other tasks had higher priorities.

**Pull request merged**